File: ClassEval_T/cpp/solution/PageUtil.cpp

#include <algorithm>
#include <cmath>

#include <vector>
#include <unordered_map>
#include <string>
#include <variant>

class PageUtil {
public:
    PageUtil(const std::vector<int>& data, int page_size);

    std::vector<int> get_page(int page_number) const;

    std::unordered_map<std::string, std::variant<int, std::vector<int>, bool>> get_page_info(int page_number) const;

    std::unordered_map<std::string, std::variant<std::string, int, std::vector<int>>> search(const std::string& keyword) const;

    std::vector<int> data;
    int page_size;
    int total_items;
    int total_pages;
};

PageUtil::PageUtil(const std::vector<int>& data, int page_size)
    : data(data), page_size(page_size) {
    total_items = data.size();
    total_pages = (total_items + page_size - 1) / page_size;
}

std::vector<int> PageUtil::get_page(int page_number) const {
    if (page_number < 1 || page_number > total_pages) {
        return {};
    }

    int start_index = (page_number - 1) * page_size;
    int end_index = start_index + page_size;
    if (end_index > total_items) {
        end_index = total_items;
    }
    return std::vector<int>(data.begin() + start_index, data.begin() + end_index);
}

std::unordered_map<std::string, std::variant<int, std::vector<int>, bool>> PageUtil::get_page_info(int page_number) const {
    if (page_number < 1 || page_number > total_pages) {
        return {};
    }

    int start_index = (page_number - 1) * page_size;
    int end_index = std::min(start_index + page_size, total_items);
    std::vector<int> page_data(data.begin() + start_index, data.begin() + end_index);

    std::unordered_map<std::string, std::variant<int, std::vector<int>, bool>> page_info = {
        {"current_page", page_number},
        {"per_page", page_size},
        {"total_pages", total_pages},
        {"total_items", total_items},
        {"has_previous", page_number > 1},
        {"has_next", page_number < total_pages},
        {"data", page_data}
    };

    return page_info;
}

std::unordered_map<std::string, std::variant<std::string, int, std::vector<int>>> PageUtil::search(const std::string& keyword) const {
    std::vector<int> results;
    for (int item : data) {
        if (std::to_string(item).find(keyword) != std::string::npos) {
            results.push_back(item);
        }
    }

    int num_results = results.size();
    int num_pages = (num_results + page_size - 1) / page_size;

    std::unordered_map<std::string, std::variant<std::string, int, std::vector<int>>> search_info = {
        {"keyword", keyword},
        {"total_results", num_results},
        {"total_pages", num_pages},
        {"results", results}
    };

    return search_info;
}

File: BatchTestTool/java/java automated tester/Test/compile_passed/solutions/PageUtil.java
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.Collections;

public class PageUtil {
    private List<Object> data;
    private int pageSize;
    private int totalItems;
    private int totalPages;

    public PageUtil(List<Object> data, int pageSize) {
        this.data = data;
        this.pageSize = pageSize;
        this.totalItems = data.size();
        this.totalPages = (totalItems + pageSize - 1) / pageSize;
    }

    public Map<String, Object> getPageInfo(int pageNumber) {
        if (pageNumber < 1 || pageNumber > totalPages) {
            return Collections.emptyMap();
        }

        int startIndex = (pageNumber - 1) * pageSize;
        int endIndex = Math.min(startIndex + pageSize, totalItems);
        List<Object> pageData = data.subList(startIndex, endIndex);

        Map<String, Object> pageInfo = new HashMap<>();
        pageInfo.put("current_page", pageNumber);
        pageInfo.put("per_page", pageSize);
        pageInfo.put("total_pages", totalPages);
        pageInfo.put("total_items", totalItems);
        pageInfo.put("has_previous", pageNumber > 1);
        pageInfo.put("has_next", pageNumber < totalPages);
        pageInfo.put("data", pageData);

        return pageInfo;
    }

    public SearchInfo search(String keyword) {
        List<Object> results = new ArrayList<>();
        for (Object item : data) {
            if (item.toString().contains(keyword)) {
                results.add(item);
            }
        }

        int numResults = results.size();
        int numPages = (numResults + pageSize - 1) / pageSize;

        SearchInfo searchInfo = new SearchInfo();
        searchInfo.setKeyword(keyword);
        searchInfo.setTotalResults(numResults);
        searchInfo.setTotalPages(numPages);
        searchInfo.setResults(results);

        return searchInfo;
    }

    public List<Object> getPage(int pageNumber) {
        if (pageNumber < 1 || pageNumber > totalPages) {
            return Collections.emptyList();
        }

        int startIndex = (pageNumber - 1) * pageSize;
        int endIndex = Math.min(startIndex + pageSize, totalItems);
        return data.subList(startIndex, endIndex);
    }

    public static class SearchInfo {
        private String keyword;
        private int totalResults;
        private int totalPages;
        private List<Object> results;

        public String getKeyword() {
            return keyword;
        }

        public void setKeyword(String keyword) {
            this.keyword = keyword;
        }

        public int getTotalResults() {
            return totalResults;
        }

        public void setTotalResults(int totalResults) {
            this.totalResults = totalResults;
        }

        public int getTotalPages() {
            return totalPages;
        }

        public void setTotalPages(int totalPages) {
            this.totalPages = totalPages;
        }

        public List<Object> getResults() {
            return results;
        }

        public void setResults(List<Object> results) {
            this.results = results;
        }
    }
}